Description: http_h2.c: properly detect when we're done with the connection
 (the client has sent GOAWAY)
Author: Ken Murchison <murch@fastmail.com>
Origin: upstream, https://github.com/cyrusimap/cyrus-imapd/commit/2696ed7
Bug: https://github.com/cyrusimap/cyrus-imapd/issues/3180
Bug-Debian: https://bugs.debian.org/969222
Forwarded: not-needed
Reviewed-By: Xavier Guimard <yadd@debian.org>
Last-Update: 2020-10-25

--- a/imap/http_h2.c
+++ b/imap/http_h2.c
@@ -653,6 +653,12 @@
             txn->flags.conn = CONN_CLOSE;
         }
     }
+    else if (!nghttp2_session_want_read(ctx->session)) {
+        /* We're done */
+        syslog(LOG_DEBUG, "closing connection");
+        txn->flags.conn = CONN_CLOSE;
+        return;
+    }
 }
 
 
@@ -662,9 +668,10 @@
     int want_read = nghttp2_session_want_read(ctx->session);
     int goaway = txn->flags.conn & CONN_CLOSE;
     nghttp2_error_code err = goaway ? NGHTTP2_REFUSED_STREAM : NGHTTP2_NO_ERROR;
+    struct protstream *pin = txn->conn->pin;
 
     syslog(LOG_DEBUG, "http2_input()  goaway: %d, eof: %d, want read: %d",
-           goaway, txn->conn->pin->eof, want_read);
+           goaway, prot_IS_EOF(pin), want_read);
 
     if (want_read && !goaway) {
         /* Read frame(s) */
@@ -682,12 +689,12 @@
         else {
             /* Failure */
             syslog(LOG_DEBUG, "nghttp2_session_recv: %s (%s)",
-                   nghttp2_strerror(r), prot_error(txn->conn->pin));
+                   nghttp2_strerror(r), prot_error(pin));
             goaway = 1;
 
             if (r == NGHTTP2_ERR_CALLBACK_FAILURE) {
                 /* Client timeout */
-                txn->error.desc = prot_error(txn->conn->pin);
+                txn->error.desc = prot_error(pin);
                 err = NGHTTP2_REFUSED_STREAM;
             }
             else {
